File: webapp/src/main/scala/org/linnando/mixemulator/webapp/MemoryDataComponent.scala
package org.linnando.mixemulator.webapp

import angulate2.std._
import org.linnando.mixemulator.vm.VirtualMachine
import org.linnando.mixemulator.vm.io.data.IOWord

import scala.collection.immutable.NumericRange
import scala.scalajs.js

@Component(
  selector = "memory-data",
  templateUrl = "webapp/src/main/resources/memory-data.component.html",
  styleUrls = @@@("webapp/src/main/resources/memory-data.component.css")
)
class MemoryDataComponent(virtualMachineService: VirtualMachineService) {
  val addresses: js.Array[Short] = js.Array[Short]() ++ NumericRange[Short](0, VirtualMachine.MEMORY_SIZE, 1)

  private def machineState = virtualMachineService.machineState

  def cellSign(address: Short): String =
    cellContent(address).map(w => if (w.negative) "-" else "+").getOrElse("+")

  def cellByte(address: Short, pos: Int): Byte =
    cellContent(address).map(_.bytes(pos)).getOrElse(0)

  private def cellContent(address: Short): Option[IOWord] = machineState.map(_.get(address))
}

File: webapp/src/app/memory-data/memory-data.component.ts
import { Component } from '@angular/core';
import { IOWord, IOWord$, MEMORY_SIZE, VirtualMachineState } from '@mixemulator/lib';
import { VirtualMachineService } from '../virtual-machine.service';

@Component({
  selector: 'mix-memory-data',
  templateUrl: './memory-data.component.html',
  styleUrls: ['./memory-data.component.scss']
})
export class MemoryDataComponent {
  readonly addresses: number[] = [...Array(MEMORY_SIZE).keys()];

  constructor(private virtualMachineService: VirtualMachineService) {
  }

  get machineState(): VirtualMachineState | undefined {
    return this.virtualMachineService.machineState;
  }

  cellSign(address: number): string {
    return this.cellContent(address).negative ? '-' : '+';
  }

  cellByte(address: number, pos: number): number {
    return this.cellContent(address).bytes[pos];
  }

  cellContent(address: number): IOWord {
    return this.machineState?.get(address) ?? IOWord$.empty();
  }
};
